src/server/utils/watchLocalModules.js

```diff
@@ -14,52 +14,145 @@
  * permissions and limitations under the License.
  */
 
-import fs from 'node:fs';
 import path from 'node:path';
-import chokidar from 'chokidar';
+import fs from 'node:fs/promises';
+import { glob } from 'glob';
 import loadModule from 'holocron/loadModule.node';
 import {
   getModules,
   getModuleMap,
   resetModuleRegistry,
   addHigherOrderComponent,
 } from 'holocron/moduleRegistry';
-import { getIp } from './getIP';
 import onModuleLoad from './onModuleLoad';
 
+const CHANGE_WATCHER_INTERVAL = 1000;
+const WRITING_FINISH_WATCHER_TIMEOUT = 400;
+
+// why our own code instead of something like https://www.npmjs.com/package/chokidar?
+// we did, but saw misses, especially when using @americanexpress/one-app-runner
+// * occasional misses even on the native filesystem (e.g. macOS)
+// * https://forums.docker.com/t/file-system-watch-does-not-work-with-mounted-volumes/12038
+// * runner on macOS chokidar would see the first change but not subsequent changes, presumably due
+//   to inode tracking and the way the build directories are destroyed and recreated
+// * using runner with --envVars.CHOKIDAR_USEPOLLING=true had no effect
+// for us the pattern of the files we want to check have a very consistent pattern and we don't need
+// to know the changes in the sometimes very many other files, so we can skip a lot of work from a
+// general solution
+
+async function changeHandler(changedPath) {
+  // I had apprehension of this manual manipulation
+  // but from what I can tell a file or directory with '/' in the name is not valid
+  // Linux: https://stackoverflow.com/q/9847288
+  // macOS: '/' from Finder is translated to ':'
+  // Windows: https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions
+  const [moduleName] = changedPath.split(path.sep);
+  const moduleMap = getModuleMap();
+  const moduleMapEntry = moduleMap.getIn(['modules', moduleName]);
+
+  if (!moduleMapEntry) {
+    console.warn(`module "${moduleName}" not in the module map, make sure to serve-module first`);
+    return;
+  }
+
+  console.log(`the Node.js bundle for ${moduleName} finished saving, attempting to load`);
+
+  let newModule;
+  try {
+    // FIXME: this leads to a race condition later
```

Comment:
I'm not sure there is a race condition

You get, edit, and reset the module map below without any await, so yes, the modules will race to load. But regardless the edits to the module map will be atomic?

Reply:
different copies are created if two modules are loaded at the same time